### DISCUSSION TITLE
2007/11/27/%E5%A4%9A%E6%80%81vi

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#192] [Announcements]
# 2007/11/27/%E5%A4%9A%E6%80%81vi

一、多态VI的概念

https://blog.qizhen.xyz/2007/11/27/%E5%A4%9A%E6%80%81vi

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Qizhen
2007年12月3日 15:04 编辑
那个 map container 后来更新过，主要是修改了一些bug。已经可以生成EXE了。新的版本只能在8.5下使用。下载它的地址没有变 http://www.ni.com/labs/哈希表，我们考虑过，但是没有尝试实现，感觉上它比二叉树要难，主要没有一个各种情况都适用的哈希算法。对于某一个特定问题，倒是可以考虑做个哈希表的。

回复
Qizhen
2007年12月3日 14:58 编辑
关于你说的只更改程序的部分VI的问题，我想一个比DLL更简单的办法是，你把可能需要更新的VI都提出来，作成动态调用的VI。每次生成的EXE中是不包含这些VI的。它们独立于可执行文件之外。需要更新，则把新的VI给用户，替换掉旧的就可以。如果VI需要保密，可以加密码，甚至保存为不含程序框图的VI，这样用户无论如何都看不到他的代码。动态调用的VI，最主要缺点是加载比较慢，一个解决办法是每次程序启动时，就把所有可能用到的VI都装入内存。

回复
Qizhen
2007年12月3日 14:53 编辑
关于嵌套，你说的没错，多态VI只是不能直接作为另外一个多态VI的实例VI，间接再调用是可以的。实例VI中再使用别的多态VI是可以的。但是，实例中再想调用自己所在的多态VI是不允许的。

回复
Qizhen
2007年12月3日 14:49 编辑
开发多态VI要稍微多费一些事，但会方便使用它的人。对与同样数据类型的输入，不同的算法，也可以做成多态VI。LabVIEW 中也确实有这样的例子。但是，这种多态VI在使用时就不能根据数据类型来自动选择实例VI了，必须要用户手工指定，这样一来，它给用户带来的好处就大打折扣了。比如说，只使用普通的VI，但是增加一个枚举类型的参数用来指定算法，也可以达到同样的目的。

回复
杰
2007年12月1日 23:53 编辑
by the way，你说很多人都会问你一些数据采集方面的问题，可能你自己没有把自己介绍清楚，或者说是NI给人的印象。因为我们大多数打交道的都是AE或者销售。他们几乎都是什么都知道一点的，呵呵。如果别人知道你是NI的R&D，估计就不会让你这么烦了。我一开始也一位你是一个AE呢。所以你可以在blog的显眼的位置注明自己是NI的R&D。或许会少点麻烦。
我来继续纠缠你一个问题。
象我现在写LV的程序，发布给用户之后，如果需要更改或者用户提出一个新的要求，那么我可能整个程序都需要重新编译再发布给用户。很不方便。虽然我可能只是改了一个子VI的一小部分，或者只是改了一个算法。
那么，能不能在我的程序就不用更改的情况下，实现用户的新的需求呢。其实也就是接口－插件技术在我的程序里面的体现。我现在的做法是把不同的功能模块做成dll文件，然后去调用。需要更改的时候，给用户发一个新的dll，覆盖原来的库文件，程序是不变的。甚至于我对我的程序的许可权（lisence）的控制等等，都在dll里面实现。一年到期了，让程序就不能工作，嘿嘿。因为dll文件不可见，所以有它的优势。
不过这么做也非常的烦人。而且根本上来说，我需要重新编译我的dll文件。有没有更好的实现的方法呢？
我想过，是不是LV里面的所有的子VI的调用都用动态调用的方式就OK了？没太整明白。还请赐教！
 
你做的map containner，最近进展如何？我用的还是比较爽，就是只能在编辑状态下用。现在有新的版本出来了吗？可以编译成exe使用了吗？
我和你交流过你的map是用的二叉树搜索算法的，有一个预先的排序，所以速度很快。然后我自己也尝试着做了一个使用链地址的哈希搜索的算法，不是很完善。没太多时间，水平也不太灵光，呵呵。

回复
杰
2007年12月1日 23:17 编辑
相当感谢！呵呵。
我已经搞出来具体怎么弄这个东西了。我理解，对于一个单独的程序而言，做成多态VI可能会比较浪费一点时间。其实你一个一个调实例VI就可以了。改起来也比较方便。不过，如果是后面有很多类似的项目，基本上会用到相同功能的VI时，前面花点时间做个多态的，会很方便。
你上方所述，大部分都是在说相对于不同的数据类型。不过我经常遇到的倒不是这个主要问题。经常遇到的是，输入的数据不仅数据类型，甚至就是同样的数据，但是我需要不同的处理的输出的数据，而且不同的处理出来的结果的数据类型也是一样的。
在有的情况下，可能是输入的数据多一路干涉（多一个connector)，处理过程多一个，但是和少一个干涉的处理基本类似。输出是一样的要求。
那么，这，也应该是多态VI的范畴吧？
在NI提供的范例VI里面，很多这样的多态VI。呵呵。在多态VI的输入端有一个选择器，进行选择之后，这个VI会自动的变成你需要的连接方式。象数据采集里面提供的DAQ卡的许多范例VI都是这样的。当然，这些多态VI也都是由实例VI来组成的。点进去看程序框图会很清楚的看到。
按照你前面的介绍，是不是就能够实现这样的功能，我得试试看，呵呵。
另外还有一个问题，你说，多态VI不能嵌套使用，一个多态VI不能作为其他多态VI的实例VI。那么假设是我做的几个很复杂的实例VI里面，调用了一个我自己做的多态VI，我能不能把这几个复杂的实例VI做成一个多态的呢？
或者简单说，多态VI只是不能直接作为另外一个多态VI的实例VI，还是只要实例VI中使用了自己做的多态VI，就不能再重复的做多态VI了？
谢谢！呵呵。
 
 

回复
